server/leafnode_test.go

```diff
@@ -1554,3 +1557,95 @@ func TestLeafNodeTLSVerifyAndMap(t *testing.T) {
 		})
 	}
 }
+
+func TestLeafNodeOriginClusterInfo(t *testing.T) {
+	hopts := DefaultOptions()
+	hopts.ServerName = "hub"
+	hopts.LeafNode.Port = -1
+
+	hub := RunServer(hopts)
+	defer hub.Shutdown()
+
+	conf := createConfFile(t, []byte(fmt.Sprintf(`
+		port: -1
+		leaf {
+			remotes [ { url: "nats://127.0.0.1:%d" } ]
+		}
+	`, hopts.LeafNode.Port)))
+
+	defer os.Remove(conf)
+	opts, err := ProcessConfigFile(conf)
+	if err != nil {
+		t.Fatalf("Error processing config file: %v", err)
+	}
+	opts.NoLog, opts.NoSigs = true, true
+
+	s := RunServer(opts)
+	defer s.Shutdown()
+
+	checkLeafNodeConnected(t, s)
+
+	// Check the info on the leadnode client in the hub.
+	grabLeaf := func() *client {
+		var l *client
+		hub.mu.Lock()
+		for _, l = range hub.leafs {
+			break
+		}
+		hub.mu.Unlock()
+		return l
+	}
+
+	l := grabLeaf()
+	if rc := l.remoteCluster(); rc != "" {
+		t.Fatalf("Expected an empty remote cluster, got %q", rc)
+	}
+
+	s.Shutdown()
+
+	// Now make our leafnode part of a cluster.
+	conf = createConfFile(t, []byte(fmt.Sprintf(`
+		port: -1
+		leaf {
+			remotes [ { url: "nats://127.0.0.1:%d" } ]
+		}
+		cluster {
+			name: "abc"
+			listen: "127.0.0.1:-1"
+		}
+	`, hopts.LeafNode.Port)))
+
+	defer os.Remove(conf)
+	opts, err = ProcessConfigFile(conf)
+	if err != nil {
+		t.Fatalf("Error processing config file: %v", err)
+	}
+	opts.NoLog, opts.NoSigs = true, true
+
+	s = RunServer(opts)
+	defer s.Shutdown()
+
+	checkLeafNodeConnected(t, s)
+
+	l = grabLeaf()
+	if rc := l.remoteCluster(); rc != "abc" {
+		t.Fatalf("Expected a remote cluster name of \"abc\", got %q", rc)
+	}
+	pcid := l.cid
+
+	// Now make sure that if we update our cluster name, simulating the settling
+	// of dynamic cluster names between competing servers.
+	s.setClusterName("xyz")
+	// Make sure we disconnect and reconnect.
+	checkLeafNodeConnectedCount(t, s, 0)
+	checkLeafNodeConnected(t, s)
+
+	l = grabLeaf()
+	if rc := l.remoteCluster(); rc != "xyz" {
```

Comment:
Getting a panic here sometimes running the test, because `l` is nil.

Reply:
https://github.com/nats-io/nats-server/pull/1493/checks?check_run_id=811791046#step:6:1206

Reply:
Should be fixed.

Reply:
Yes, that's same issue in test that I had in the past (but in my case it did reveal a real underlying bug).